[filesystem][next] Add options to `downloadFileAsync` function

LeonDvlpmnt opened the pull request:
# Why

I was migrating our application to filesystem/next when I noticed it wasn't possible to set headers anymore for downloading files. Because all our downloads are only downloadable with an API key we always send an authentication header along with the download. 

I could probably use `expo/fetch` nowadays as well to achieve this but I wanted to challenge myself to implement it for filesystem instead :)

# How

Implemented `DownloadOptionsNext` on the native side. For now it only adds headers to `OkHttpClient` and `URLSession`. In the future other options can be implemented as well.

# Test Plan

Added a basic suite test. Also tested locally with a simple node API that requires a certain header.

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

LeonDvlpmnt committed:
Add options to download function

LeonDvlpmnt committed:
Add options to download function

LeonDvlpmnt committed:
Update packages/expo-file-system/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

LeonDvlpmnt committed:
Update packages/expo-file-system/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

LeonDvlpmnt committed:
lint fix

LeonDvlpmnt committed:
lint fix

LeonDvlpmnt committed:
prettier fixes

LeonDvlpmnt committed:
prettier fixes

LeonDvlpmnt committed:
rebuild

LeonDvlpmnt committed:
rebuild

leonhh commented:
@aleqsio any chance you could take a look at this?

leonhh commented:
Test failure seems to be unrelated

leonhh commented:
@brentvatne @aleqsio @jakex7 sorry for the ping but is there any chance this could get reviewed? I would love to fully migrate to filesystem/next in SDK 53

aleqsio reviewed (approved):
Looks good, apologies for missing this earlier on!

aleqsio commented:
Don't worry about the failing test – if you can rebase and rebuild it happy to land this!

lukmccall committed:
[modules-plugin] Fix CI

LeonDvlpmnt committed:
rebuild

lukmccall committed:
[modules-plugin][Android] Handle case where `shouldUsePublicationScript` is null

LeonDvlpmnt committed:
revert submodule change

gabrieldonadel committed:
[build-properties] Add enableBundleCompression option (#36071)

# Why

Closes ENG-15456

# How

Add `android.enableBundleCompression` option to expo-build-properties
and updates bare-minimum template

# Test Plan


```sh
$ npx create-expo-app --template path/to/expo/expo/templates/expo-template-blank sdk53
$ cd sdk53 
$ yarn add file:/path/to/expo/expo/packages/expo-build-properties
# update `app.json` plugin config (Add ``android.enableBundleCompression`)
$ expo prebuild --template /path/to/expo/expo/expo-template-bare-minimum-*.tgz
```

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

---------

Co-authored-by: Kudo Chien <kudo@expo.dev>

Simek committed:
[docs] APISection: improve rendering literal types in props (#36072)

Simek committed:
[cli] Readme: formatting, updated E2E instructions (#34633)

leonhh commented:
@aleqsio @jakex7 thanks for the review, all done!

behenate committed:
[expo][2/n] Roll out edge-to-edge: Add `react-native-edge-to-edge` as a dependency of `expo` (#35812)

# Why

We need all projects to support configuring the edge to edge feature in
SDK53. We use `react-native-edge-to-edge` for that, so we add is as a
dependency of `expo`.
It is not an expo module so it won't be autolinked as it is a transitive
dependency for the project. I have added a special case to autolinking,
where we check if the `expo` version is greater than `52` and if it is
`react-native-edge-to-edge` is added to the list of autolinked packages.

# How

Added `react-native-edge-to-edge` as a dependency of the `expo` package
Added a special case in autolinking that checks if the expo version is
high enough and adds the `react-native-edge-to-edge` to autolinking.

# Test Plan

Tested in a seperate project, which had `edge-to-edge` added as a
transitive dependency via a random 3rd party package, and patched
autolinking to contain code from this PR (with a change of expo version
from 53 to 52 so that the `if` gets triggered). Edge-to-edge has
installed and autolinked successfuly.
Also worked for BareExpo and Expo Go.

brentvatne committed:
[docs] Add 53 beta docs

marklawlor committed:
[router]: Add prefetch functionality (#34558)

# Why

Support React Navigation 7 route preloading

# How

There are three ways to enable preloading

- `<Link />` component with the `preload` attribute will preload the
requested Href when the Link's screen is focused.
- `router.preload(<href>)`. The imperative API to `preload` a route.
- `<Preload href={<href>} />` The preload component will reload the
passed Href when the component's screen is focused.


I initially implemented this with new logic to handle groups/nested
navigators, but then I realised that we can re-use the `linkTo`
function, which greatly simplifies the code and correctly targets the
right navigator. The main advantage is being able to reload routes for
other navigators and correctly preloading routes for deeply nested
screens.


Documentation will be handled in a seperate PR. We should note that
preloading routes is handled by the navigators and not guaranteed to
work for all types of navigators.

# Test Plan

Added unit tests. Tested in simulator.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

---------

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

Simek committed:
[docs] fix build after defining new beta, add missing data file (#36077)

Kudo committed:
[doc] add sqlite web experimental callout (#36078)

# Why

add note for expo-sqlite web is still experimental

# How

add an experimental callout

behenate committed:
[expo][3/n] Roll out edge-to-edge: Add `enableEdgeToEdge` app config field (#35958)

# Why

We want users to be able to configure `edge-to-edge` from the app config
file.

# How

We use `react-native-edge-to-edge` for enabling `edge-to-edge`.
If `enableEdgeToEdge` is false or undefined
- We warn about a future drop of support
- Disable the force-enable for Android 15
If `enableEdgeToEdge` is true
- Import `react-native-edge-to-edge` config plugin and apply it with
some reasonable defaults (DayNight theme)
If someone configured `react-native-edge-to-edge` plugin directly
- Apply the plugin as normal

# Test Plan

Tested in a test project created in `expo/apps` 
I also tested interaction with `androidStatusBar` properties, and they
work as expected (`android:statusBarColor`) doesn't work in edge-to-edge
mode.

Simek committed:
[expo] add missing entries and versions for bundled packages (#36080)

# Why

Spotted that we miss version information about few packages in native
bundled modules data.

# How

Add `@expo/fingerprint`, `expo-manifests`, `expo-mesh-gradient` and
`expo-live-photo` to the bundled native modules list.

# Test Plan

Package checks are passing locally.

marklawlor committed:
[server]: Change redirects to occur before rewrites (#35931)

# Why

Swap rewrites and redirects so they cannot chain

betomoedano committed:
[docs] Add Turso to local first (#36082)

# Why

[ENG-15419 Update local-first doc with
Turso](https://linear.app/expo/issue/ENG-15419/update-local-first-doc-with-turso)

# Test Plan

Ran locally

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

---------

Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

EvanBacon committed:
feat(babel): Update React Compiler support for React 19. (#36081)

# Why

- SDK 53 uses React 19 so we should update React Compiler to target it.
This is technically breaking but compiler is marked as beta.
- Updated the docs to indicate the simpler setup in React 19

# Test Plan

- Tested in a beta project.

---------

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

marklawlor committed:
[router]: Remove git diff conflict commnet

EvanBacon committed:
chore(cli): deprecate --https flag (#36083)

# Why

- The `--https` flag generates an insecure certificate which is fragile,
heavy, doesn't work well across computers, and breaks in Expo Go.
- Using `https://localhost` is only partially supported in many tools,
e.g. Apple auth does not support localhost at all.
- `https` used to be required for even basic web APIs such as crypto but
these often work fine with `http://localhost`, reducing the need for the
insecure method.
- A better alternative is to use the free tunneling service we provide
via `expo start --tunnel` to generate a temporary hosted https URL with
proper certificates that work across all platforms and enable
authentication development.

# How

- Mark `--https` flag as deprecated in the docs and CLI.
- Update docs to inform users about the `--tunnel` flag instead of
`--https`.

---------

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

brentvatne committed:
[expo-ui] Fix application of scrollEnabled prop

Kudo committed:
[ui][ncl] add SwiftUI container example (#35555)

# Why

add a NCL example to demonstrate swiftui without uiview integration
close ENG-15236

behenate committed:
[expo][4/n] Roll out edge-to-edge: Enable by default in new projects (#35959)

# Why

Since edge-to-edge is always in enforced in Android 16+ we want to
enable edgeToEdge by default in new projects to have as many people
onboard as possible.

# How

Updated the template files.

**should we also update the minimal template?**
Please let me know if there is anything else that has to be updated to
change the templates.

# Test Plan

Ran `npx create-expo-app -t <local-path-to-template-directory>`
I'm not sure if there is a better way of testing this

brentvatne committed:
Publish packages

uri-scheme@1.4.2
expo-web-browser@14.1.2
@expo/ui@0.1.1-alpha.1
expo-task-manager@13.1.2
expo-sqlite@15.2.3
expo-splash-screen@0.30.2
expo-router@5.0.2-preview.2
expo-modules-core@2.3.3
expo-modules-autolinking@2.1.3
expo-file-system@18.1.2
expo-build-properties@0.14.2
expo-background-task@0.2.3
expo-background-fetch@13.1.2
expo@53.0.0-preview.4
create-expo-nightly@0.1.1
babel-preset-expo@13.1.2
@expo/server@0.6.2
@expo/prebuild-config@8.1.3
@expo/config-types@53.0.0-preview.2
@expo/cli@0.23.4

brentvatne committed:
[templates] Update to latest [skip ci]

brentvatne committed:
[babel-preset-expo] Update compiler snapshot

Kudo committed:
[babel-preset] throw error for import.meta by default (#36074)

# Why

follow up #34756

# How

add a default transform error if people having `import.meta` but not
enable unstable_transformImportMeta

---------

Co-authored-by: Brent Vatne <brentvatne@gmail.com>
Co-authored-by: Evan Bacon <bacon@expo.io>

chrsmys committed:
[secure-store] Add support for iOS application groups (#36056)

This adds support for access groups in iOS keychain so that items can be
shared securely across apps and app extensions.

https://developer.apple.com/documentation/security/sharing-access-to-keychain-items-among-a-collection-of-apps

softwarebyze committed:
[docs] change "`HomeScreen` class" to "`HomeScreen` component" (#36084)

teppeita committed:
[docs] Fix outdated links in AppDelegateSubscribers documentation (#36090)

clarkg committed:
[ios][contacts][permissions] Disambiguate Contacts permissions between granted and limited access on iOS (#35772)

alanjhughes committed:
[skip ci] Bump expo go version

Simek committed:
[build-properties] few tweaks for JSDoc comments (#36094)

# Why

Supersedes #34526 

# How

Add changes from mentioned above PR, plus few other tweaks for JSDoc
comments.

# Test Plan

![Screenshot 2025-04-11 at 12 39
29](https://github.com/user-attachments/assets/4301cbd1-0e84-4e9f-ae73-b5f124531b39)

aleqsio committed:
[expo-file-system][next] Fix inconsistent behavior when using special chars in filenames. (#35801)

Simek committed:
[docs] handle beta docs correctly in version schema fetcher (#36095)

# Why

Spotted that after adding new reference docs beta version we still were
displaying SDK 53 bundled version in Next docs.

# How

Take beta version in account when fetching data for Next docs,
regenerate the data files stored in repository.

# Test Plan

While running docs app locally, the bundled version is no longer visible
on the Next reference pages, beta pages include the correct information,
packages like Fingerprint or Manifests also lists now the bundled
version correctly in beta docs.

# Preview

![Screenshot 2025-04-11 at 12 57
29](https://github.com/user-attachments/assets/0dc1a287-d50f-4984-9a35-13b275bb5b41)

alanjhughes committed:
[skip ci] Publish home

aleqsio committed:
Change `registerWebModule` to work with minified classes (#35946)

Co-authored-by: Kudo Chien <kudo@expo.dev>

gabrieldonadel committed:
chore(cli): Use eas-cli fingerprint:generate to generate fingerprints (#36085)

# Why

To ensure that locally calculated fingerprints match `eas build`
fingerprints, we should use the same configurations. Utilizing `eas-cli
fingerprint:generate` directly simplifies keeping these configurations
in sync. Additionally, when we calculate fingerprints with eas cli, the
sources get automatically uploaded.

# How

Attempt to use `eas fingerprint:generate` to check for compatible builds
when remove build cache provider is `eas`

# Test Plan

Run `nexpo run:android` and `nexpo run:ios` locally

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

Simek committed:
[docs] Snippet: more compact actions on mobile, overflow tweaks (#36096)

# Why

Fixes #31618 

# How

Add more compact variant for snippet actions on mobile, overflow tweaks
for title/header of a snippet.

# Test Plan

The changes have been reviewed by running docs app locally, and visiting
few pages which use code snippet component.

All lint checks and tests are passing.

# Preview

![Screenshot 2025-04-11 at 14 20
48](https://github.com/user-attachments/assets/8c77741c-a5e6-4da2-a90a-43083d0f8f0d)

![Screenshot 2025-04-11 at 14 22
38](https://github.com/user-attachments/assets/20955cd3-b2c8-4390-a35d-b621c42127c7)

![Screenshot 2025-04-11 at 14 23
01](https://github.com/user-attachments/assets/1a6c6c4a-ff8f-401d-9ab8-23b567ca60c9)

jakex7 committed:
[cli] Output recommended versions in prebuild (#35941)

# Why

Resolves ENG-15404

# How

Updates CLI output.

# Test Plan

Run `yarn expo prebuild` with different from recommended versions (eg in
bare-expo)
 
| Before | After |
|---|---|
|
![image](https://github.com/user-attachments/assets/f0c37552-8340-4b67-b52b-b27abb85f508)
|
![image](https://github.com/user-attachments/assets/defb8e71-d458-4912-b285-1fb9d695c1c3)
|

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

---------

Co-authored-by: Brent Vatne <brentvatne@gmail.com>

gabrieldonadel committed:
chore(cli): Warn users to upgrade eas-cli version (#36103)

# Why

For the remote builds cache feature to work, users have to be running
the latest version of eas-cli

# How

Add a warning for users to upgrade their eas-cli

# Test Plan

Run `nexpo run:ios`

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

brentvatne committed:
[status-bar] Use react-native-edge-to-edge system bars when edge-to-edge is enabled (#36087)

# Why

To make the migration to edge-to-edge easier

# How

There is actually not a lot going on here, it just looks like a big
change because I had to shuffle the files around.

-
[types.ts](https://github.com/expo/expo/blob/brent/eng-15467-make-expo-status-bar-use-systembars-when-edge-to-edge-is/packages/expo-status-bar/src/types.ts)
and
[NativeStatusBarWrapper.tsx](https://github.com/expo/expo/blob/brent/eng-15467-make-expo-status-bar-use-systembars-when-edge-to-edge-is/packages/expo-status-bar/src/NativeStatusBarWrapper.tsx)
- I split these out from the former main StatusBar component file. I
wanted to be able to isolate any Android specific logic around
edge-to-edge to an `.android.tsx` file, but still re-use the normal
implementation when edge-to-edge is not enabled.
-
[StatusBar.android.tsx](https://github.com/expo/expo/blob/brent/eng-15467-make-expo-status-bar-use-systembars-when-edge-to-edge-is/packages/expo-status-bar/src/StatusBar.android.tsx)
- this imports `NativeStatusBarWrapper` and wraps each of its methods,
so if edge-to-edge is enabled it just calls those directly, otherwise it
now has different behavior.
- Tests had to be re-organized a bit.
- Otherwise everything is basically the same.

LeonDvlpmnt committed:
[router] bump @radix-ui/react-slot (#36089)

# Why

Fixes peer dependency on react 18.3.1. This change will make sure react
19 is used instead.

# How

Bump @radix-ui/react-slot to 1.2.0

Before:
```
  "peerDependencies": {
    "react": "^16.8 || ^17.0 || ^18.0"
  },
```
After:
```
  "peerDependencies": {
    "@types/react": "*",
    "react": "^16.8 || ^17.0 || ^18.0 || ^19.0 || ^19.0.0-rc"
  },
```

# Test Plan

Tests ran successfully,<Slot /> works in the sandbox app, haven't tested
asChild.

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

kadikraman committed:
[eslint-config-expo] fix browser globals for flat config (#36101)

byCedric committed:
fix(metro-config): `withSerializerPlugins` maintaining original config object reference (#36099)

**TL;DR;** `withSerializerPlugins` clones the Metro config object,
losing the original config reference - and by that - losing changes made
in `metro.config.js` - affecting/breaking any changes to
`getModulesRunBeforeMainModule` for exports only.

# Why

During the Expo Go prerelease, we ran into a weird issue only with
production exports - development (`expo start`) builds were working
fine. Every export resulted in the error:


![image](https://github.com/user-attachments/assets/4b74caaf-d47c-469d-bb41-613572a0f9c5)

Usually, this error points in the direction of
`getModulesRunBeforeMainModule` not including
`react-native/Libraries/Core/InitializeCore.js` - as this sets up XHR
(including `FormData`) global polyfills. But Atlas pointed out that this
is indeed imported in the dependency chain. Only after running `expo
export -p web --no-bytecode --no-minify` it was clear that the
`getModulesRunBeforeMainModule` wasn't serialized properly:

What it should have | What it had
--- | ---

![image](https://github.com/user-attachments/assets/d4336254-554e-4593-a5cc-47467d950978)
|
![image](https://github.com/user-attachments/assets/948ba4f2-5226-4574-a6ce-c3010ad87a86)

In Expo Go, we [modify the main
modules](https://github.com/expo/expo/blob/0293c79ae07f3f37e6d900063424956652f2de32/apps/expo-go/metro.config.js#L57-L71)
to point to our Expo Go-specific React Native fork - something we didn't
used to do before (which was a mistake).

After debugging this intensively with break points and clicking through,
it turns out that the serializer config did not include the changes made
in `metro.config.js` - instead the `getModulesRunBeforeMainModule`
contained the default function. This is because we actually clone the
whole config object when creating the serializers - which means that any
changes in `metro.config.js` are NOT propagated as it lost the object
reference to the main config.

This fixes that.

# How

- Instead of cloning, mutate the Metro config with the serializer to
maintain the original config object
- This fixes the `getModulesRunBeforeMainModule` changes we need in Expo
Go

# Test Plan

See added test

### Test with Expo Go

This Metro config has a modified `getModulesRunBeforeMainModule`
already.

- `cd react-native-lab/react-native`
- `yarn install --frozen-lockfile`
- `cd apps/expo-go`
- `yarn expo export -p ios --no-bytecode --no-minify`
- `tail dist/_expo/static/js/ios/index-<hash>.js`
  - Should contain 2 `__r(<moduleId>)` statements
  - The first one is the `react-native/Libraries/Core/InitializeCore.js`
  - The second one is the app entry point

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources
according to [this short
guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

LeonDvlpmnt committed:
Merge branch 'feat/allow-setting-headers-for-downloads' of github.com:leonhh/expo into feat/allow-setting-headers-for-downloads
